File: packages/monosecret/test/monosecret_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:monosecret/monosecret.dart';
import 'package:test/test.dart';

void main() {
  group('MonosecretClient.get', () {
    test('returns stdout with trailing CLI whitespace removed', () async {
      final cli = await _fakeCli(r'''
#!/usr/bin/env sh
printf 'secret-value  \n\n'
''');

      final client = MonosecretClient(executable: cli.path);

      expect(await client.get('API_KEY'), 'secret-value');
    });

    test(
      'passes secret name and optional profile/provider/file flags',
      () async {
        final dir = await Directory.systemTemp.createTemp(
          'monosecret_dart_test_',
        );
        addTearDown(() => dir.delete(recursive: true));

        final argsFile = File('${dir.path}/args.json');
        final cli = await _recordingCli(argsFile: argsFile, stdout: 'value\n');
        final client = MonosecretClient(executable: cli.path);

        await client.get(
          'DATABASE_URL',
          profile: 'production',
          provider: 'op',
          file: 'monosecret.production.toml',
        );

        expect(await _readRecordedArgs(argsFile), [
          'get',
          'DATABASE_URL',
          '--profile',
          'production',
          '--provider',
          'op',
          '--file',
          'monosecret.production.toml',
        ]);
      },
    );
  });

  group('MonosecretClient.check', () {
    test('uses --no-prompt by default and forwards selectors', () async {
      final dir = await Directory.systemTemp.createTemp(
        'monosecret_dart_test_',
      );
      addTearDown(() => dir.delete(recursive: true));

      final argsFile = File('${dir.path}/args.json');
      final cli = await _recordingCli(argsFile: argsFile);
      final client = MonosecretClient(executable: cli.path);

      await client.check(
        profile: 'ci',
        provider: 'env',
        file: 'monosecret.ci.toml',
      );

      expect(await _readRecordedArgs(argsFile), [
        'check',
        '--no-prompt',
        '--profile',
        'ci',
        '--provider',
        'env',
        '--file',
        'monosecret.ci.toml',
      ]);
    });

    test('omits --no-prompt when disabled', () async {
      final dir = await Directory.systemTemp.createTemp(
        'monosecret_dart_test_',
      );
      addTearDown(() => dir.delete(recursive: true));

      final argsFile = File('${dir.path}/args.json');
      final cli = await _recordingCli(argsFile: argsFile);
      final client = MonosecretClient(executable: cli.path);

      await client.check(noPrompt: false);

      expect(await _readRecordedArgs(argsFile), ['check']);
    });
  });

  group('MonosecretClient.loadEnvironment', () {
    test(
      'passes include/group selectors and returns injected environment',
      () async {
        final dir = await Directory.systemTemp.createTemp(
          'monosecret_dart_test_',
        );
        addTearDown(() => dir.delete(recursive: true));

        final argsFile = File('${dir.path}/args.json');
        final cli = await _environmentCli(
          argsFile: argsFile,
          injectedEnvironment: {
            'DATABASE_URL': 'postgres://localhost/app',
            'API_KEY': 'abc123',
          },
        );
        final client = MonosecretClient(executable: cli.path);

        final environment = await client.loadEnvironment(
          include: ['DATABASE_URL', 'API_KEY'],
          groups: ['backend', 'workers'],
          profile: 'development',
          provider: 'dotenv',
          file: 'monosecret.toml',
        );

        expect(
          environment,
          containsPair('DATABASE_URL', 'postgres://localhost/app'),
        );
        expect(environment, containsPair('API_KEY', 'abc123'));
        expect(await _readRecordedArgs(argsFile), [
          'run',
          '--profile',
          'development',
          '--provider',
          'dotenv',
          '--file',
          'monosecret.toml',
          '--include',
          'DATABASE_URL',
          '--include',
          'API_KEY',
          '--group',
          'backend',
          '--group',
          'workers',
          '--',
          if (Platform.isWindows) ...['cmd', '/c', 'set'] else ...['env'],
        ]);
      },
    );

    test(
      'parses environment lines and preserves equals signs in values',
      () async {
        final cli = await _fakeCli('''
#!/usr/bin/env sh
printf 'DATABASE_URL=postgres://localhost/app\nTOKEN=value=with=equals\nIGNORED_LINE\n'
''');
        final client = MonosecretClient(executable: cli.path);

        expect(await client.loadEnvironment(), {
          'DATABASE_URL': 'postgres://localhost/app',
          'TOKEN': 'value=with=equals',
        });
      },
    );
  });

  group('process configuration', () {
    test('passes custom environment to the CLI process', () async {
      final cli = await _fakeCli(r'''
#!/usr/bin/env sh
printf '%s' "$MONOSECRET_TEST_TOKEN"
''');
      final client = MonosecretClient(
        executable: cli.path,
        environment: {'MONOSECRET_TEST_TOKEN': 'from-client-env'},
      );

      expect(await client.get('TOKEN'), 'from-client-env');
    });

    test('runs the CLI from the configured working directory', () async {
      final dir = await Directory.systemTemp.createTemp(
        'monosecret_dart_test_',
      );
      addTearDown(() => dir.delete(recursive: true));

      final cli = await _fakeCli(r'''
#!/usr/bin/env sh
pwd
''');
      final client = MonosecretClient(
        executable: cli.path,
        workingDirectory: dir.path,
      );

      expect(
        await File(await client.get('PWD')).resolveSymbolicLinks(),
        await dir.resolveSymbolicLinks(),
      );
    });
  });

  group('MonosecretException', () {
    test('captures command, exit code, stdout, stderr, and message', () async {
      final cli = await _fakeCli(r'''
#!/usr/bin/env sh
echo "partial output"
echo "boom" >&2
exit 7
''');
      final client = MonosecretClient(executable: cli.path);

      await expectLater(
        client.check(profile: 'ci'),
        throwsA(
          isA<MonosecretException>()
              .having((error) => error.args, 'args', [
                'check',
                '--no-prompt',
                '--profile',
                'ci',
              ])
              .having((error) => error.exitCode, 'exitCode', 7)
              .having((error) => error.stdout, 'stdout', 'partial output\n')
              .having((error) => error.stderr, 'stderr', 'boom\n')
              .having(
                (error) => error.toString(),
                'toString',
                contains(
                  'monosecret check --no-prompt --profile ci failed with exit code 7: boom',
                ),
              ),
        ),
      );
    });
  });
}

Future<File> _fakeCli(String source) async {
  final dir = await Directory.systemTemp.createTemp('monosecret_dart_test_');
  addTearDown(() => dir.delete(recursive: true));

  final file = File('${dir.path}/monosecret');
  await file.writeAsString(source);
  await _makeExecutable(file);
  return file;
}

Future<File> _recordingCli({required File argsFile, String stdout = ''}) {
  return _fakeCli('''
#!/usr/bin/env sh
${_writeArgsSnippet(argsFile.path)}
printf '%s' ${shellQuote(stdout)}
''');
}

Future<File> _environmentCli({
  required File argsFile,
  required Map<String, String> injectedEnvironment,
}) {
  final exports = injectedEnvironment.entries
      .map((entry) => 'export ${entry.key}=${shellQuote(entry.value)}')
      .join('\n');

  return _fakeCli('''
#!/usr/bin/env sh
${_writeArgsSnippet(argsFile.path)}
while [ "\$#" -gt 0 ]; do
  if [ "\$1" = "--" ]; then
    shift
    break
  fi
  shift
done
$exports
exec "\$@"
''');
}

String _writeArgsSnippet(String path) {
  return '''
ARGS_FILE=${shellQuote(path)}
: > "\$ARGS_FILE"
for arg do
  printf '%s\n' "\$arg" >> "\$ARGS_FILE"
done
''';
}

Future<List<String>> _readRecordedArgs(File file) async {
  return const LineSplitter().convert(await file.readAsString());
}

Future<void> _makeExecutable(File file) async {
  final result = await Process.run('chmod', ['+x', file.path]);
  if (result.exitCode != 0) {
    throw StateError('chmod failed: ${result.stderr}');
  }
}

String shellQuote(String value) => "'${value.replaceAll("'", "'\\''")}'";

File: npm/monosecret__client/test/monosecret.test.ts
import { mkdtemp, readFile, realpath, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { MonosecretClient, MonosecretException, parseEnvironment } from "../src/index.js";

const tempDirs: string[] = [];

const nodeClient = (
  script: string,
  options: Omit<
    ConstructorParameters<typeof MonosecretClient>[0],
    "executable" | "executableArgs"
  > = {},
) =>
  new MonosecretClient({
    ...options,
    executable: process.execPath,
    executableArgs: [script],
  });

describe("MonosecretClient.get", () => {
  it("returns stdout with trailing CLI whitespace removed", async () => {
    const cli = await fakeCli({ stdout: "secret-value  \n\n" });
    const client = nodeClient(cli);

    await expect(client.get("API_KEY")).resolves.toBe("secret-value");
  });

  it("passes secret name and optional profile/provider/file flags", async () => {
    const argsFile = await tempFile("args.json");
    const cli = await fakeCli({ argsFile, stdout: "value\n" });
    const client = nodeClient(cli);

    await client.get("DATABASE_URL", {
      file: "monosecret.production.toml",
      profile: "production",
      provider: "op",
    });

    await expect(readRecordedArgs(argsFile)).resolves.toEqual([
      "get",
      "DATABASE_URL",
      "--profile",
      "production",
      "--provider",
      "op",
      "--file",
      "monosecret.production.toml",
    ]);
  });
});

describe("MonosecretClient.check", () => {
  it("uses --no-prompt by default and forwards selectors", async () => {
    const argsFile = await tempFile("args.json");
    const cli = await fakeCli({ argsFile });
    const client = nodeClient(cli);

    await client.check({
      file: "monosecret.ci.toml",
      profile: "ci",
      provider: "env",
    });

    await expect(readRecordedArgs(argsFile)).resolves.toEqual([
      "check",
      "--no-prompt",
      "--profile",
      "ci",
      "--provider",
      "env",
      "--file",
      "monosecret.ci.toml",
    ]);
  });

  it("omits --no-prompt when disabled", async () => {
    const argsFile = await tempFile("args.json");
    const cli = await fakeCli({ argsFile });
    const client = nodeClient(cli);

    await client.check({ noPrompt: false });

    await expect(readRecordedArgs(argsFile)).resolves.toEqual(["check"]);
  });
});

describe("MonosecretClient.loadEnvironment", () => {
  it("passes include/group selectors and returns injected environment", async () => {
    const argsFile = await tempFile("args.json");
    const cli = await environmentCli({
      argsFile,
      injectedEnvironment: {
        API_KEY: "abc123",
        DATABASE_URL: "postgres://localhost/app",
      },
    });
    const client = nodeClient(cli);

    const environment = await client.loadEnvironment({
      file: "monosecret.toml",
      groups: ["backend", "workers"],
      include: ["DATABASE_URL", "API_KEY"],
      profile: "development",
      provider: "dotenv",
    });

    expect(environment).toMatchObject({
      API_KEY: "abc123",
      DATABASE_URL: "postgres://localhost/app",
    });
    await expect(readRecordedArgs(argsFile)).resolves.toEqual([
      "run",
      "--profile",
      "development",
      "--provider",
      "dotenv",
      "--file",
      "monosecret.toml",
      "--include",
      "DATABASE_URL",
      "--include",
      "API_KEY",
      "--group",
      "backend",
      "--group",
      "workers",
      "--",
      ...(process.platform === "win32" ? ["cmd", "/c", "set"] : ["env"]),
    ]);
  });

  it("parses environment lines and preserves equals signs in values", () => {
    expect(
      parseEnvironment(
        "DATABASE_URL=postgres://localhost/app\nTOKEN=value=with=equals\nIGNORED_LINE\n",
      ),
    ).toEqual({
      DATABASE_URL: "postgres://localhost/app",
      TOKEN: "value=with=equals",
    });
  });
});

describe("process configuration", () => {
  it("passes custom environment to the CLI process", async () => {
    const cli = await fakeCli({ stdoutEnvironmentVariable: "MONOSECRET_TEST_TOKEN" });
    const client = nodeClient(cli, {
      environment: { MONOSECRET_TEST_TOKEN: "from-client-env" },
    });

    await expect(client.get("TOKEN")).resolves.toBe("from-client-env");
  });

  it("runs the CLI from the configured working directory", async () => {
    const dir = await tempDir();
    const cli = await fakeCli({ stdoutWorkingDirectory: true });
    const client = nodeClient(cli, { workingDirectory: dir });

    await expect(client.get("PWD")).resolves.toBe(await realpath(dir));
  });
});

describe("MonosecretException", () => {
  it("captures command, exit code, stdout, stderr, and message", async () => {
    const cli = await fakeCli({ exitCode: 7, stderr: "boom\n", stdout: "partial output\n" });
    const client = nodeClient(cli);

    await expect(client.check({ profile: "ci" })).rejects.toMatchObject({
      args: ["check", "--no-prompt", "--profile", "ci"],
      exitCode: 7,
      name: "MonosecretException",
      stderr: "boom\n",
      stdout: "partial output\n",
    } satisfies Partial<MonosecretException>);
    await expect(client.check({ profile: "ci" })).rejects.toThrow(
      `${process.execPath} check --no-prompt --profile ci failed with exit code 7: boom`,
    );
  });
});

afterEach(async () => {
  await Promise.all(tempDirs.splice(0).map((dir) => rm(dir, { force: true, recursive: true })));
});

async function fakeCli(
  options: {
    argsFile?: string;
    exitCode?: number;
    stderr?: string;
    stdout?: string;
    stdoutEnvironmentVariable?: string;
    stdoutWorkingDirectory?: boolean;
  } = {},
): Promise<string> {
  return writeCli(`
    import { writeFileSync } from 'node:fs';

    const args = process.argv.slice(2);
    ${options.argsFile === undefined ? "" : `writeFileSync(${JSON.stringify(options.argsFile)}, JSON.stringify(args));`}
    ${options.stdoutWorkingDirectory === true ? "process.stdout.write(process.cwd());" : ""}
    ${options.stdoutEnvironmentVariable === undefined ? "" : `process.stdout.write(process.env[${JSON.stringify(options.stdoutEnvironmentVariable)}] ?? '');`}
    ${options.stdout === undefined ? "" : `process.stdout.write(${JSON.stringify(options.stdout)});`}
    ${options.stderr === undefined ? "" : `process.stderr.write(${JSON.stringify(options.stderr)});`}
    process.exit(${options.exitCode ?? 0});
  `);
}

async function environmentCli(options: {
  argsFile: string;
  injectedEnvironment: Record<string, string>;
}): Promise<string> {
  const stdout = Object.entries(options.injectedEnvironment)
    .map(([key, value]) => `${key}=${value}`)
    .join("\n");

  return fakeCli({
    argsFile: options.argsFile,
    stdout: `${stdout}\n`,
  });
}

async function writeCli(source: string): Promise<string> {
  const file = await tempFile("monosecret.mjs");
  await writeFile(file, source);
  return file;
}

async function readRecordedArgs(file: string): Promise<string[]> {
  return JSON.parse(await readFile(file, "utf8")) as string[];
}

async function tempFile(name: string): Promise<string> {
  return join(await tempDir(), name);
}

async function tempDir(): Promise<string> {
  const dir = await mkdtemp(join(tmpdir(), "monosecret_ts_test_"));
  tempDirs.push(dir);
  return dir;
}
